suprb/optimizer/rule/ns/novelty_calculation.py

```diff
@@ -0,0 +1,100 @@
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+from suprb.rule import Rule
+from suprb.base import BaseComponent
+from scipy.spatial.distance import hamming
+from .novelty_search_type import NoveltySearchType
+from .archive import Archive
+
+
+class NoveltyCalculation(BaseComponent, metaclass=ABCMeta):
+    def __init__(self, novelty_search_type: NoveltySearchType, archive: Archive, k_neighbor: int):
+        self.novelty_search_type = novelty_search_type
+        self.archive = archive
+        self.k_neighbor = k_neighbor
+
+    def __call__(self, rules: list[Rule]) -> list[Rule]:
+        return self._novelty_score(rules=rules)
+
+    def _add_distances_to_archive_rules(self, rule: Rule, archive: list[Rule]):
+        for i, _ in enumerate(archive):
+            archive[i].distance = hamming(rule.match_, archive[i].match_)
```

Comment:
should we make the distance function configurable?

Reply:
Only if you already have other distance functions in mind (as I would like to add them to the smoke test). Otherwise I would keep it like this and if we decide to try out different distance function we make it configurable

Reply:
Jaccard could be a candidate. If negative values are no issue, Dice as well. There might be others for which I could check again.

Reply:
Created a new issue for this: https://github.com/heidmic/suprb/issues/122